akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala

```diff
@@ -148,6 +132,104 @@ private[http] object OutgoingConnectionBlueprint {
+          case ResponseStart(statusCode, protocol, headers, entityCreator, _) ⇒
+            val entity = createEntity(entityCreator) withSizeLimit parserSettings.maxContentLength
+            push(out, HttpResponse(statusCode, headers, entity, protocol))
+
+          case MessageStartError(_, info) ⇒
+            throw IllegalResponseException(info)
+
+          case other ⇒
+            throw new IllegalStateException(s"ResponseStart expected but $other received.")
+        }
+
+        def onPull(): Unit = {
+          if (!entitySubstreamStarted) pull(in)
+        }
+      }
+
+      private val waitForMessageEnd = new InHandler {
+        def onPush(): Unit = grab(in) match {
+          case MessageEnd ⇒ setHandler(in, idle)
+          case other      ⇒ throw new IllegalStateException(s"MessageEnd expected but $other received.")
+        }
+      }
+
+      setHandler(in, idle)
+      setHandler(out, idle)
+
+      private def createEntity(creator: EntityCreator[ResponseOutput, ResponseEntity]): ResponseEntity = {
+        creator match {
+          case StrictEntityCreator(entity) ⇒
+            pull(in)
+            setHandler(in, waitForMessageEnd)
+            entity
+
+          case StreamedEntityCreator(creator) ⇒
+            val entitySource = new SubSourceOutlet[ResponseOutput]("EntitySource")
+            entitySubstreamStarted = true
+            var responseCancelled = false
+
+            entitySource.setHandler(new OutHandler {
+              override def onPull(): Unit = pull(in)
+
+              override def onDownstreamFinish(): Unit = {
+                responseCancelled = true
+                if (isClosed(in) || eagerClose) {
+                  entitySource.complete()
+                  completeStage()
+                } else {
+                  // Start draining to prevent stalling
+                  if (!hasBeenPulled(in)) pull(in)
+                }
+              }
+            })
+
+            setHandler(in, new InHandler {
+              override def onPush(): Unit = grab(in) match {
+                case MessageEnd ⇒
+                  entitySource.complete()
+                  entitySubstreamStarted = false
+                  responseCancelled = false
+                  setHandler(in, idle)
+                  setHandler(out, idle)
+
+                case messagePart ⇒
+                  // Drain response to prevent stalling
+                  if (responseCancelled) pull(in)
+                  else entitySource.push(messagePart)
+              }
+
+              override def onUpstreamFinish(): Unit = {
+                entitySource.complete()
+                completeStage()
+              }
+
+              override def onUpstreamFailure(reason: Throwable): Unit = {
+                entitySource.fail(reason)
+                failStage(reason)
+              }
+            })
+
+            creator(Source.fromGraph(entitySource.source))
+        }
+      }
+    }
+  }
```

Comment:
Please give it another quick look, I moved things around a bit to avoid unnecessary allocations.